binary_search_tree/tree.py

```diff
@@ -7,49 +10,135 @@ def __init__(self, key, val = None):
+                return
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def add(self, key, value = None):
-        pass
+        new_node = TreeNode(key, value)
+        if self.root == None:
+            self.root = new_node
+        else:
+            self.add_helper(self.root, new_node)
+
+    def find_helper(self, node, key):
+        if node == None:
+            return None
+        if node.key == key:
+            return node.value
+        if key < node.key:
+            return self.find_helper(node.left, key)
+        else:
+            return self.find_helper(node.right, key)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        return self.find_helper(self.root, key)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
+    def inorder_helper(self, node, arr):
+        if node == None:
+            return
+        self.inorder_helper(node.left, arr)
+        arr.append({"key": node.key, "value": node.value})
+        self.inorder_helper(node.right, arr)
 
-    # Time Complexity: 
-    # Space Complexity: 
     def inorder(self):
-        pass
+        arr = []
+
+        self.inorder_helper(self.root, arr)
+        return arr
+
+    def preorder_helper(self, node, arr):
+        if node == None:
+            return
+        
+        arr.append({"key": node.key, "value": node.value})
+        self.preorder_helper(node.left, arr)
+        self.preorder_helper(node.right, arr)
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def preorder(self):
-        pass
+        arr = []
+
+        self.preorder_helper(self.root, arr)
+        return arr
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def postorder_helper(self, node, arr):
+        if node == None:
+            return
+        
+        self.postorder_helper(node.left, arr)
+        self.postorder_helper(node.right, arr)
+        arr.append({"key": node.key, "value": node.value})
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def postorder(self):
-        pass
+        arr = []
+        self.postorder_helper(self.root, arr)
+        return arr
 
-    # Time Complexity: 
-    # Space Complexity:     
-    def height(self):
-        pass
+    def height_helper(self, node):
+        if not node:
+            return 0
+        else:
+            return 1 + max(self.height_helper(node.left), self.height_helper(node.right))
 
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)  
+    def height(self):
+        return self.height_helper(self.root)
 
 #   # Optional Method
-#   # Time Complexity: 
-#   # Space Complexity: 
+#   # Time Complexity: O(n^2) (due to using a list & deleting from the front of the list, could be O(n) with a better queue implementation
+#   # Space Complexity: O(n)
     def bfs(self):
```

Comment:
👍 Since you visit each node twice (when you add it to the Queue and when you pop it and add it's children), the time complexity is O(n)